File: android/app/src/main/kotlin/com/timecoin/wallet/crypto/Keypair.kt
package com.timecoin.wallet.crypto

import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer
import java.security.SecureRandom

/**
 * Ed25519 keypair — wraps BouncyCastle to match the desktop wallet's
 * ed25519-dalek Keypair (generate, sign, verify).
 */
class Keypair private constructor(
    private val privateKey: Ed25519PrivateKeyParameters,
    val publicKey: Ed25519PublicKeyParameters,
) {
    /** 32-byte public key */
    fun publicKeyBytes(): ByteArray = publicKey.encoded

    /** 32-byte secret key */
    fun secretKeyBytes(): ByteArray = privateKey.encoded

    /** Hex-encoded secret key */
    fun secretKeyHex(): String = secretKeyBytes().toHexString()

    /** Sign a message, returning a 64-byte Ed25519 signature. */
    fun sign(message: ByteArray): ByteArray {
        val signer = Ed25519Signer()
        signer.init(true, privateKey)
        signer.update(message, 0, message.size)
        return signer.generateSignature()
    }

    /** Verify a signature against this keypair's public key. */
    fun verify(message: ByteArray, signature: ByteArray): Boolean {
        return verifyWithPublicKey(publicKeyBytes(), message, signature)
    }

    companion object {
        /** Generate a new random keypair. */
        fun generate(): Keypair {
            val random = SecureRandom()
            val privateKey = Ed25519PrivateKeyParameters(random)
            val publicKey = privateKey.generatePublicKey()
            return Keypair(privateKey, publicKey)
        }

        /** Create a keypair from a 32-byte secret key. */
        fun fromBytes(secretBytes: ByteArray): Keypair {
            require(secretBytes.size == 32) { "Secret key must be 32 bytes" }
            val privateKey = Ed25519PrivateKeyParameters(secretBytes, 0)
            val publicKey = privateKey.generatePublicKey()
            return Keypair(privateKey, publicKey)
        }

        /** Create a keypair from a hex-encoded secret key. */
        fun fromHex(hex: String): Keypair {
            val bytes = hex.hexToByteArray()
            return fromBytes(bytes)
        }

        /** Verify a signature using a raw 32-byte public key. */
        fun verifyWithPublicKey(
            publicKey: ByteArray,
            message: ByteArray,
            signature: ByteArray,
        ): Boolean {
            require(publicKey.size == 32) { "Public key must be 32 bytes" }
            require(signature.size == 64) { "Signature must be 64 bytes" }
            val pubKeyParams = Ed25519PublicKeyParameters(publicKey, 0)
            val verifier = Ed25519Signer()
            verifier.init(false, pubKeyParams)
            verifier.update(message, 0, message.size)
            return verifier.verifySignature(signature)
        }
    }
}

// Hex utilities
fun ByteArray.toHexString(): String = joinToString("") { "%02x".format(it) }

fun String.hexToByteArray(): ByteArray {
    require(length % 2 == 0) { "Hex string must have even length" }
    return chunked(2).map { it.toInt(16).toByte() }.toByteArray()
}

File: ios/TimeCoinWallet/Crypto/Keypair.swift
import Foundation
import CryptoKit

// Ed25519 keypair — wraps CryptoKit, matches BouncyCastle Ed25519 (RFC 8032).
struct Keypair {
    let privateKey: Curve25519.Signing.PrivateKey
    var publicKey: Curve25519.Signing.PublicKey { privateKey.publicKey }

    var publicKeyBytes: Data { publicKey.rawRepresentation }
    var secretKeyBytes: Data { privateKey.rawRepresentation }
    var secretKeyHex: String { secretKeyBytes.hexString }

    static func generate() -> Keypair {
        Keypair(privateKey: Curve25519.Signing.PrivateKey())
    }

    static func fromBytes(_ secretBytes: Data) throws -> Keypair {
        guard secretBytes.count == 32 else {
            throw KeypairError.invalidSecretKeyLength(secretBytes.count)
        }
        let key = try Curve25519.Signing.PrivateKey(rawRepresentation: secretBytes)
        return Keypair(privateKey: key)
    }

    static func fromHex(_ hex: String) throws -> Keypair {
        let bytes = try Data(hexString: hex)
        return try fromBytes(bytes)
    }

    func sign(_ message: Data) throws -> Data {
        try privateKey.signature(for: message)
    }

    func verify(message: Data, signature: Data) -> Bool {
        publicKey.isValidSignature(signature, for: message)
    }

    static func verifyWithPublicKey(_ publicKeyBytes: Data, message: Data, signature: Data) -> Bool {
        guard let pubKey = try? Curve25519.Signing.PublicKey(rawRepresentation: publicKeyBytes) else {
            return false
        }
        return pubKey.isValidSignature(signature, for: message)
    }
}

enum KeypairError: Error, LocalizedError {
    case invalidSecretKeyLength(Int)
    case invalidHex

    var errorDescription: String? {
        switch self {
        case .invalidSecretKeyLength(let l): return "Secret key must be 32 bytes, got \(l)"
        case .invalidHex: return "Invalid hex string"
        }
    }
}

// MARK: - Hex utilities

extension Data {
    var hexString: String {
        map { String(format: "%02x", $0) }.joined()
    }

    init(hexString: String) throws {
        guard hexString.count % 2 == 0 else {
            throw KeypairError.invalidHex
        }
        var data = Data()
        var index = hexString.startIndex
        while index < hexString.endIndex {
            let next = hexString.index(index, offsetBy: 2)
            guard let byte = UInt8(hexString[index..<next], radix: 16) else {
                throw KeypairError.invalidHex
            }
            data.append(byte)
            index = next
        }
        self = data
    }
}
